src/connector/src/sink/kafka.rs

```diff
@@ -441,32 +462,104 @@ impl KafkaSinkWriter {
             }
         }
 
+        if commit_flag {
+            // Give the buffer back to the origin
+            std::mem::swap(&mut future_buffer, &mut self.future_delivery_buffer);
+
+            // Sanity check
+            debug_assert!(
+                future_buffer.is_empty(),
+                "future buffer must be empty after swapping"
+            );
+
+            match self.commit().await {
+                // FIXME: Is this error handling enough?
+                Ok(_) => return Ok(()),
+                Err(_) => return Err(err),
+            }
+        }
+
+        if push_flag {
+            // Indicates success
+            std::mem::swap(&mut future_buffer, &mut self.future_delivery_buffer);
+
+            // Sanity check
+            debug_assert!(
+                future_buffer.is_empty(),
+                "future buffer must be empty after swapping"
+            );
+
+            return Ok(());
+        }
+
         Err(err)
     }
 
     async fn write_json_objects(
-        &self,
+        &mut self,
         event_key_object: Option<Value>,
         event_object: Option<Value>,
     ) -> Result<()> {
+        let topic = self.config.common.topic.clone();
         // here we assume the key part always exists and value part is optional.
         // if value is None, we will skip the payload part.
         let key_str = event_key_object.unwrap().to_string();
-        let mut record = FutureRecord::<[u8], [u8]>::to(self.config.common.topic.as_str())
-            .key(key_str.as_bytes());
+        let mut record = FutureRecord::<[u8], [u8]>::to(topic.as_str()).key(key_str.as_bytes());
         let payload;
         if let Some(value) = event_object {
             payload = value.to_string();
             record = record.payload(payload.as_bytes());
         }
+        // Send the data but not wait it to finish sinking
+        // Will join all `DeliveryFuture` during commit
         self.send_result(record).await?;
         Ok(())
     }
 
-    async fn debezium_update(&self, chunk: StreamChunk, ts_ms: u64) -> Result<()> {
+    async fn commit(&mut self) -> Result<()> {
+        // Get the ownership first
```

Comment:
The code can be simplified into the following
```
try_join_all(self.future_delivery_buffer.drain().map(|kafka_result| kafka_result.and_then(|kafka_result| kafka_result))).await?
```

Reply:
IIUC, there are two possible flaws with this approach.
1. We can not distinguish the error type (i.e., Whether the error is due to producer or the failure during the delivery)
2. The `and_then` will produce `Result<(i32, i64), (KafkaError, OwnedMessage)>` which is essentially not a future to be awaited, and `try_join_all` will produce a `Vec` containing all the successful results, which also is not compatible with the original return type.

Thus I'd prefer the current logic at present.

Reply:
The general idea of the code snippet above is to first turn the return type of the future from nested result `Result<Result<(i32, i64), (KafkaError, OwnedMessage)>, Canceled>` to `Result<(i32, i64), SinkError>`, and then do a `try_join_all` on the iterator of transformed future, which is the same as the current logic, but in a more elegant way.

The return type of future is changed by via `map` on the future. The `and_then` is applied inside the closure of `map`, which is turning the nested result into a plain result.

Reply:
I guess below is the simplification you'd prefer. 
(Tried many times, this is the most elegant solution I could come up with)
```rust
async fn commit_inner(&mut self) -> Result<()> {
        let _v = try_join_all(
            self.future_delivery_buffer
                .drain(..)
                .map(|x| {
                    x.map(|f| {
                        match f {
                            Ok(Ok(val)) => Ok(val),
                            Ok(Err((k_err, _msg))) => Err(SinkError::Kafka(k_err)),
                            Err(_) => Err(SinkError::Kafka(KafkaError::Canceled)),
                        }
                    })
                })
        )
        .await?;

        Ok(())
    }
```

In addition, I'd prefer readability rather than elegant but hard-to-follow code snippet🤔️.

Reply:
The code snippet LGTM, and I think the code is intuitive and not hard-to-follow. All involved methods are commonly used library method. 

`try_join_all` is a commonly used util method for rust async users, when we want to join some result future and either collect all ok result or stop when seeing any err. The current for-loop is a total reimplementation of it, or even a worse one, because if any future at the queue end completes with an err, we should stop polling other future at the queue front and return the error, but current for loop should wait for the completion of future at the queue front before seeing the err at the end. `drain` is also commonly used when we want to consume all items in a vector without taking the ownership of it. 

For code readability, the code is hard to read in current code snippet since the parameter of closure is meaningless `x` and `f`. After we rename `x` to `future` and `f` to `future_result`, it will become much more readable.

One more thing we can improve in the code snippet above is that, since we don't care about value returned from the result, when we transform the return type of future, we'd better turn it into `Ok(())`. This is because the return type of `try_join_all` is a vector, and a vector of `val` involves unnecessary memory allocation. But `()` is zero-sized, and `Vec<()>` will not involve any memory allocation.

Reply:
Could we use `VecDeque` to replace `Vec` ? I means that we only need to control the length of this queue. If the head of future in queue is ready, we can pull more request, rather than waiting all request in queue finished.